Logs performance benchmarks

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Improves usability and realism of the performance tests for logging.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/assign @serathius

pohly committed:
test/integration/logs: update benchmark support

When trying again with recent log files from the CI job, it was found that some
JSON messages get split across multiple lines, both in container logs and in
the systemd journal:

   2022-12-21T07:09:47.914739996Z stderr F {"ts":1671606587914.691,"caller":"rest/request.go:1169","msg":"Response ...
   2022-12-21T07:09:47.914984628Z stderr F 70 72  6f 78 79 10 01 1a 13 53 ... \".|\n","v":8}

Note the different time stamp on the second line. That first line is
long (17384 bytes). This seems to happen because the data must pass through a
stream-oriented pipe and thus may get split up by the Linux kernel.

The implication is that lines must get merged whenever the JSON decoder
encounters an incomplete line. The benchmark loader now supports that. To
simplifies this, stripping the non-JSON line prefixes must be done before using
a log as test data.

The updated README explains how to do that when downloading a CI job
result. The amount of manual work gets reduced by committing symlinks under
data to the expected location under ci-kubernetes-kind-e2e-json-logging and
ignoring them when the data is not there.

Support for symlinks gets removed and path/filepath is used instead of path
because it has better Windows support.

pohly committed:
test/integration/logs: remove useless stats case

The same effect can be achieved with `-bench=BenchmarkEncoding/none`.

pohly committed:
test/integration/logs: benchmark using logsapi

The benchmarks and unit tests were written so that they used custom APIs for
each log format. This made them less realistic because there were subtle
differences between the benchmark and a real Kubernetes component. Now all
logging configuration is done with the official
k8s.io/component-base/logs/api/v1.

To make the different test cases more comparable, "messages/s" is now reported
instead of the generic "ns/op".

pohly commented:
@serathius wrote in https://github.com/kubernetes/kubernetes/pull/115277#issuecomment-1406265369:
> Overall benchmark changes makes sense for me, however I'm not sure why we need the symlink magic. Could you explain more about why you added it?

With the directories set up like this, running benchmarks against some log files from the CI only need a single command as setup (the gsutil mentioned in the updated README). I could keep the symlinks out of the commit, but creating them or moving files is complicated enough that I then would have to add a script instead that users need to invoke - I don't think that this will be simpler.

Without any symlinks, the same file would have to be replicated in different places (once for each verbosity level).

pohly commented:
Hmm, I had forgotten about the additional mangling of the files. Perhaps a script is better after all.

serathius commented on test/integration/logs/benchmark/get-logs.sh:
```diff
@@ -0,0 +1,54 @@
+grep 'kind-worker kubelet' ci-kubernetes-kind-e2e-json-logging/artifacts/kind-worker/kubelet.log | \
+    sed -e 's;^.* kind-worker kubelet[^ ]*: ;;' > ci-kubernetes-kind-e2e-json-logging/kind-worker-kubelet.log
+
+ln -s ../ci-kubernetes-kind-e2e-json-logging/kind-worker-kubelet.log data/kind-worker-kubelet.log
```
Why `ln -s` and not `mv`? What's the benefit of using symlinks?

pohly replied:
The same file gets referenced in two different directories and then serves as input for different test cases (default log level and v 3). I could copy it, but that seems wasteful.

pohly replied:
@serathius: I've added some comments to the script about this and also completed the support for running the script multiple times (would have failed when symlinks already exist).

pohly commented:
/retest

serathius commented:
Not super convinced on necessity of symlinks in repo. Would love some second opinion. @dims

dims commented:
looks like this is a developer oriented script right? let's leave it simple (and have them duplicated)

pohly commented:
> Not super convinced on necessity of symlinks in repo.

I already removed them?

logicalhan commented:
/triage accepted

pohly commented:
@dims, @serathius: what exactly do you want me to change?

pohly committed:
test/integration/logs: replace assert.Contains

For long strings the output of assert.Contains is not very readable.

pohly committed:
test/integration/logs: use stable struct for unit test

v1.Container is still changing a log which caused the test to fail each time a
new field was added. To test loading, let's better use something that is
unlikely to change. The runtimev1.VersionResponse gets logged by kubelet and
seems to be stable.

pohly commented:
> Not super convinced on necessity of symlinks in repo.

I replaced with plain copies.

There was a merge conflict because `v1.Container` was modified by some different PR in the meantime. To avoid having to update the `load_test.go` each time that happens, I added one commit where I picked some different struct to test with.

@serathius: please take another look.

serathius commented on test/integration/logs/benchmark/.gitignore:
```diff
@@ -0,0 +1,11 @@
+# The following files resp. symlinks are created by get-logs.sh.
```
Is this still needed?

pohly replied:
No, simplified.

serathius commented on test/integration/logs/benchmark/benchmark_test.go:
```diff
@@ -50,11 +52,23 @@ func BenchmarkEncoding(b *testing.B) {
 		messages, stats, err := loadLog(path)
+		if err != nil {
+			if os.IsNotExist(err) {
+				// Ignore symlinks to missing files.
```
Is this still needed as we removed symlinks?

pohly replied:
No, removed.

serathius commented on test/integration/logs/benchmark/get-logs.sh:
```diff
@@ -0,0 +1,65 @@
+gsutil -m cp -R "gs://kubernetes-jenkins/logs/ci-kubernetes-kind-e2e-json-logging/${job}/*" ci-kubernetes-kind-e2e-json-logging/
+
+for i in kube-apiserver kube-controller-manager kube-scheduler; do
+    sed -e 's/^20[^ ]* stderr . //' \
```
One problem I expect for this script is log format changing and we forget how the original logs looked. Without knowing original log lines we will loose understanding what sed script did.

Can you provide an example what you expect this sed does? A example of line before, and example of line after

pohly replied:
Added as comments.

serathius commented on test/integration/logs/benchmark/get-logs.sh:
```diff
@@ -0,0 +1,65 @@
+done
+
+grep 'kind-worker kubelet' ci-kubernetes-kind-e2e-json-logging/artifacts/kind-worker/kubelet.log | \
+    sed -e 's;^.* kind-worker kubelet[^ ]*: ;;' > ci-kubernetes-kind-e2e-json-logging/kind-worker-kubelet.log
```
ditto

pohly replied:
Added.

pohly commented on test/integration/logs/benchmark/benchmark_test.go:
```diff
@@ -50,11 +52,23 @@ func BenchmarkEncoding(b *testing.B) {
+		}
+		// Only print file statistics once for each actual file, even if it
+		// gets linked to multiple times.
+		realPath, err := filepath.EvalSymlinks(path)
```
Hmm, now I regret that we are not symlinking anymore. Each individual copy of a file gets printed again, which can lead to substantial redundant output.

Let me replace with hashing the file content.

pohly replied:
Done. While at it, I switched from "path" to "path/filepath" because it works better on Windows.

serathius commented:
/lgtm
/approve

pohly commented:
/retest
